blockstore/splitstore/markset_badger.go

```diff
@@ -184,8 +186,53 @@ func (s *BadgerMarkSet) Has(c cid.Cid) (bool, error) {
 		return false, nil
 
 	default:
-		return false, xerrors.Errorf("error checking badger markset: %w", err)
+		return false, err
+	}
+}
+
+// writer holds the exclusive lock; put releases it
+func (s *BadgerMarkSet) put(key string) error {
+	s.pend[key] = struct{}{}
+	if len(s.pend) < badgerMarkSetBatchSize {
+		s.mx.Unlock()
+		return nil
+	}
+
+	pend := s.pend
+	seqno := s.seqno
+	s.seqno++
+	s.writing[seqno] = pend
+	s.pend = make(map[string]struct{})
+	s.mx.Unlock()
+
+	defer func() {
+		s.mx.Lock()
+		defer s.mx.Unlock()
+
+		s.version++
+		delete(s.writing, seqno)
```

Comment:
Do we really want to do this on failure?

Reply:
should we let them accumulate? It probably doesn't matter at all however, since a failure will quickly bubble up and stop the logic that is using the markset.

Reply:
It's more correct that way, right?

Reply:
I mean, we _should_ just end up failing out and not making _any_ changes. But I'd feel _slightly_ safer this way (and it can't hurt).

Reply:
it also means that we can't close and we need a separate count for writers (as it was before).
But fine, I will check the error and not remove it if we fail, and also add a writers count to disambiguate for closing.

Reply:
done.